editors/code/src/main.ts

```diff
@@ -8,15 +8,76 @@ import { activateInlayHints } from './inlay_hints';
 import { activateStatusDisplay } from './status_display';
 import { Ctx } from './ctx';
 import { Config, NIGHTLY_TAG } from './config';
-import { log, assert } from './util';
+import { log, assert, nearestParentWithCargoToml, createWorkspaceWithNewLocation } from './util';
 import { PersistentState } from './persistent_state';
 import { fetchRelease, download } from './net';
 import { spawnSync } from 'child_process';
 import { activateTaskProvider } from './tasks';
 
 let ctx: Ctx | undefined;
+const foundProjects: Set<string> = new Set();
+let config: Config | undefined = undefined;
+
+async function locateRustProjects(root: vscode.Uri) {
+    const cargoRoots = await Promise.all(vscode.workspace.textDocuments.map((doc) => nearestParentWithCargoToml(root, doc.uri)));
+    for (const cargoRoot of cargoRoots) {
+        if (cargoRoot != null) {
+            foundProjects.add(cargoRoot.fsPath);
+        }
+    }
+}
 
 export async function activate(context: vscode.ExtensionContext) {
+
+    config = new Config(context);
+
+    const workspaceFolder = vscode.workspace.workspaceFolders?.[0];
+    if (workspaceFolder !== undefined) {
+        await locateRustProjects(workspaceFolder.uri);
+    }
+
+    vscode.workspace.onDidOpenTextDocument(async (doc) => {
+        const wf = vscode.workspace.getWorkspaceFolder(doc.uri);
+        if (wf) {
+            const cargoRoot = await nearestParentWithCargoToml(wf.uri, doc.uri);
+            if (cargoRoot != null) {
+                const isMissing = !foundProjects.has(cargoRoot.fsPath);
+
+                foundProjects.add(cargoRoot.fsPath);
+                if (isMissing) {
+                    if (config?.autoRestartOnNew) {
```

Comment:
My experience in IntelliJ tells me that auto restart doesn't worth the trouble. The problem is that we should only restart in quiescent state, and only the user knows when this state is reached.

Reply:
Ie, setting a *status* "needs restart" is absolutely necessary, and a big missing piece in rust-analyzer at the moment, but it should be the user who decides to click the restart button.

Reply:
I added it after testing it a bit, since it got very old very quickly having to restart manually all the time, which is also why it's turned off by default.